Add Experience

alec016 opened the pull request:
- Added Experience Requirement(json, CT, KJS)
- Added Experience PerTick Requirement(json, CT, KJS)
- Added Experience Component
- Added Experience GuiElement (see images below)
  - Display info: ![image](https://github.com/Frinn38/Custom-Machinery/assets/78179790/5e30722c-6a1c-480c-9d15-10cd870552cc)
  - Insert Experience Levels(1, 10, all available): ![image](https://github.com/Frinn38/Custom-Machinery/assets/78179790/8702a0e4-97ae-4a3b-85c5-25bc1028c67d) ![image](https://github.com/Frinn38/Custom-Machinery/assets/78179790/7143e78d-00b1-4d21-b3e4-cca783ef6439) ![image](https://github.com/Frinn38/Custom-Machinery/assets/78179790/c6e2c0b4-f7cc-48bf-86e0-eeb43902193d)
  - Extract Experience Levels(1, 10, all available): ![image](https://github.com/Frinn38/Custom-Machinery/assets/78179790/5d5ba253-c124-4358-b0f4-0e48017c89bc) ![image](https://github.com/Frinn38/Custom-Machinery/assets/78179790/980600bf-7486-4f01-bd24-e27aa81749ae) ![image](https://github.com/Frinn38/Custom-Machinery/assets/78179790/e7bc687f-37cc-4dbd-8187-18cd4cf41c6d)

- Machine Experience Component example: `{ "type": "custommachinery:experience", "id": "xp", "capacity": 10000, "maxInput": 10000, "maxOutput": 10000, "retrieveFromSlots": true, "retrieveSlotsId": [ "output1", "input1" ] }`
- Experience GuiElement examples: 
  - Input all available: `{ "id": "xp", "x": 65, "y": 15, "type": "custommachinery:experience", "mode": "input_all" }`
  - Input 10 Levels: `{ "id": "xp", "x": 65, "y": 15, "type": "custommachinery:experience", "mode": "input_ten" }`
  - Input 1 Level: `{ "id": "xp", "x": 65, "y": 15, "type": "custommachinery:experience", "mode": "input_one" }`
  - Output all available: `{ "id": "xp", "x": 65, "y": 15, "type": "custommachinery:experience", "mode": "output_all" }`
  - Output 10 Levels: `{ "id": "xp", "x": 65, "y": 15, "type": "custommachinery:experience", "mode": "output_ten" }`
  - Output 1 Level: `{ "id": "xp", "x": 65, "y": 15, "type": "custommachinery:experience", "mode": "output_one" }`
  - Display Info:
    - Experience Points: `{ "id": "xp", "x": 113, "y": 15, "type": "custommachinery:experience", "display": "literal", "mode": "display" }`
    - Levels: `{ "id": "xp", "x": 113, "y": 15, "type": "custommachinery:experience", "display": "level", "mode": "display" }`
    - Experience Points & Levels: `{ "id": "xp", "x": 113, "y": 15, "type": "custommachinery:experience", "display": "both", "mode": "display" }`
- Experience Requirement: 
  - KubeJS & CraftTweaker: 
    - `.requireXp(float amount)`
    - `.produceXp(float amount)`
    - `.requireXpPerTick(float amount)`
    - `.produceXpPerTick(float amount)`
  - JSON:
    - ExperienceRequirement:
      - `{ "type": "custommachinery:experience", "mode": ["input", "output"], "amount": [float number], "chance": [double number between 0.0 and 1.0] }`
    - ExperienceRequirementPerTick:
      - `{ "type": "custommachinery:experience_per_tick", "mode": ["input", "output"], "amount": [float number], "chance": [double number between 0.0 and 1.0] }`
- JEI rendering ingredient: 
  - Uses GuiElement with diplay mode, otherwise dont render anything.

alec016 committed:
Add Experience Component, GuiElement, Recipe Requirement

Frinn38 commented:
That's awesome work, feature-wise it seems to be what I had in mind.
I have a few problems with your PR : 
 - Please don't use wildcard imports I avoided using them in the whole mod but your PR add some.
 - The experience component and gui element "id" property seems useless since there will only be one of this component.
 - The "maxInput" and "maxOutput" properties seems pointless too, unless I'm missing something the only ways to interact with the xp storage is with recipes which bypass the limit or with the widget which the player can just spam.
 - I don't see a point for the `IExperienceStorage` interface.
 - The experience requirement should be able to require/produce levels of xp too.

That's the observations I made for now by just browsing your PR on Github, I will do some testing as soon as I have the time.

alec016 commented:
Ok. I will check all of that and fix it.

alec016 committed:
Added level to experience requirement.
Optimized level calculation.
Remove wildcard imports.
Remove experience component "id", "maxInput", "maxOutput" properties.
Remove IExperienceStorage interface(not necessary).

alec016 committed:
Merge branch 'Frinn38:1.19.2' into 1.19.2

alec016 committed:
Changed float to int value.
Updated Codecs.fromEnum -> NamedCodec.enumCodec

alec016 committed:
Changed float to int value.
Removed maxIn maxOut properties.

alec016 committed:
Changed float to int value.

Frinn38 commented:
Finally had some time to test this, there are a few bugs that need to be fixed before merging : 
- The tooltips of the gui element in display mode are not updated when the player add or remove xp using the button or when a recipe is adding/removing xp, because that tooltip is created in the constructor of the widget the player must close and reopen the machine gui to update the tooltip (by re-creating the widget instance).
fix -> the tooltip must be dynamically created in the `ExperienceGuiElementWidget#getTooltips` method instead of the constructor.
- `ExperienceGuiElementWidget#isClickable` must return false if the gui element is in display mode as it's not a button then.
- When the player click on "add 1 level" button, 1 level is removed from the player's xp and the corresponding xp value is added to the component. This should be the other way around, the machine should take the equivalent amount of xp from the player to store 1 more level in the component (or take all xp if there isn't enough to reach the next level).
- When a recipe add x level to the machine it actually add the amount of xp needed to go from level 0 to level x. Instead it should add the amount of xp needed to go from its current level to current level + x. Same for extracting levels.
- Not really a bug but the recipe methods in `ExperienceMachineComponent` seems redundant, you could make the requirement use the other methods instead and remove the recipe specific ones.
- In the `ExperienceMachineComponent.Template` CODEC you can put the default values directly in `optionalFieldOf` methods instead of using `Optional#orElse` after.

With these few bugs fixed it should be fine to merge this PR.

If you have any questions about that feel free to ping me on discord.

alec016 committed:
- fix ExperienceGuiElementWidget#isClickable -> return false if the gui element is in display mode.
- fix ExperienceMachineComponent.Template.CODEC default values.
- fix add/extract levels when click from player level -> component level.
- fix tooltip update.
- fix recipe methods now use default methods instead of specific ones.
